src/thing_creature.c

```diff
@@ -980,337 +1102,486 @@ void first_apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx,
+{
+    if (spell_lev > SPELL_MAX_LEVEL)
+    {
+        spell_lev = SPELL_MAX_LEVEL;
+    }
+    GameTurnDelta duration = get_spell_full_duration(spell_idx, spell_lev);
+    long i = get_free_spell_slot(thing);
+    if (i != -1)
     {
-        HitPoints i = saturate_set_signed(thing->health + pwrdynst->strength[spell_lev], 16);
-        if (i < 0)
+        // Fill the spell slot if the spell has a continuous effect.
+        if (set_thing_spell_flags(thing, spell_idx, duration, spell_lev)
+        || spell_is_continuous(spell_idx, duration))
         {
-          thing->health = 0;
-        } else {
-          thing->health = min(i,cctrl->max_health);
+            fill_spell_slot(thing, spell_idx, duration, spell_lev, plyr_idx, i);
+            update_aura_effect_to_thing(thing, spell_idx);
         }
-        break;
     }
-    default:
-        break;
+    return;
+}
+
+void reapply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx, CrtrExpLevel spell_lev, PlayerNumber plyr_idx, int slot_idx)
+{
+    if (spell_lev > SPELL_MAX_LEVEL)
+    {
+        spell_lev = SPELL_MAX_LEVEL;
     }
-    if (spconf->aura_effect != 0)
+    GameTurnDelta duration = get_spell_full_duration(spell_idx, spell_lev);
+    // Reset the spell duration if the spell has a continuous effect.
+    if (set_thing_spell_flags(thing, spell_idx, duration, spell_lev)
+    || spell_is_continuous(spell_idx, duration))
     {
-        cctrl->spell_aura = spconf->aura_effect;
-        cctrl->spell_aura_duration = spconf->duration;
+        struct CreatureControl *cctrl = creature_control_get_from_thing(thing);
+        struct CastedSpellData *cspell = &cctrl->casted_spells[slot_idx];
+        cspell->duration = duration;
+        cspell->caster_level = spell_lev;
+        cspell->caster_owner = plyr_idx;
+        update_aura_effect_to_thing(thing, spell_idx);
     }
+    return;
 }
 
-void apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx, long spell_lev)
+void apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx, CrtrExpLevel spell_lev, PlayerNumber plyr_idx)
 {
-    // Make sure the creature level isn't larger than max spell level
+    struct CreatureControl *cctrl = creature_control_get_from_thing(thing);
+    if (creature_control_invalid(cctrl))
+    {
+        ERRORLOG("Invalid creature %s index %d tried to accept spell %s", thing_model_name(thing), (int)thing->index, spell_code_name(spell_idx));
+        return; // Exit the function, control is invalid.
+    }
+    struct SpellConfig *spconf = get_spell_config(spell_idx);
+    if (spell_config_is_invalid(spconf))
+    {
+        ERRORLOG("Spell %s config is invalid", spell_code_name(spell_idx));
+        return; // Exit the function, spell config is invalid.
+    }
+    // TODO: Add a check for specific spell_idx immunity that checks if creature is immune to a specific spell.
```

Comment:
You moved everything to flags, but now you want to implement the old way back too?

Reply:
Why would the new CleanseFlags not be enough?

Reply:
Sorry, the `SpellImmunity` flags on the unit configs